adservertargeting/respdataprocessor.go

```diff
@@ -0,0 +1,201 @@
+		return
+	}
+	for _, respTargetingData := range adServerTargetingData.ResponseTargetingData {
+		key := resolveKey(respTargetingData, bidderName)
+		path := respTargetingData.Path
+
+		pathSplit := strings.Split(path, pathDelimiter)
+
+		switch pathSplit[0] {
+
+		case "seatbid":
+			switch pathSplit[1] {
+			case "bid":
+				getValueFromSeatBidBid(key, path, targetingData, bidsHolder, bidderName, bid, warnings)
+			case "ext":
+				getValueFromSeatBidExt(key, path, targetingData, seatExt, warnings)
+			}
+
+		case "ext":
+			getValueFromRespExt(key, path, targetingData, response.Ext, warnings)
+		default:
+			// path points to value in response, not seat or ext
+			// bidder response has limited number of properties
+			// instead of unmarshal the whole response with seats and bids
+			// try to find data by field name
+			getValueFromResp(key, path, targetingData, response, warnings)
+
+		}
+	}
+}
+
+func buildBidExt(targetingData map[string]string,
+	bid openrtb2.Bid,
+	warnings []openrtb_ext.ExtBidderMessage,
+	truncateTargetAttribute *int) json.RawMessage {
+
+	targetingDataTruncated := truncateTargetingKeys(targetingData, truncateTargetAttribute)
+
+	bidExtTargetingData := openrtb_ext.ExtBid{
+		Prebid: &openrtb_ext.ExtBidPrebid{
+			Targeting: targetingDataTruncated,
+		},
+	}
+	bidExtTargeting, err := json.Marshal(bidExtTargetingData)
+	if err != nil {
+		warnings = append(warnings, createWarning(err.Error()))
+		return nil
+	}
+
+	newExt, err := jsonpatch.MergePatch(bid.Ext, bidExtTargeting)
+	if err != nil {
+		warnings = append(warnings, createWarning(err.Error()))
+		return nil
+	}
+	return newExt
+}
+
+func resolveKey(respTargetingData ResponseTargetingData, bidderName string) string {
+	key := respTargetingData.Key
+	if respTargetingData.HasMacro {
+		key = strings.Replace(respTargetingData.Key, bidderMacro, bidderName, -1)
+	}
+	return key
+}
+
+func truncateTargetingKeys(targetingData map[string]string, truncateTargetAttribute *int) map[string]string {
+	maxLength := MaxKeyLength
+	if truncateTargetAttribute != nil {
+		maxLength = *truncateTargetAttribute
+		if maxLength < 0 {
+			maxLength = MaxKeyLength
+		}
+	}
+
+	targetingDataTruncated := make(map[string]string)
+	for key, value := range targetingData {
+		newKey := openrtb_ext.TargetingKey(key).TruncateKey(maxLength)
+		targetingDataTruncated[newKey] = value
+	}
+	return targetingDataTruncated
+}
+
+func getValueFromSeatBidBid(
+	key, path string,
+	targetingData map[string]string,
+	bidsHolder bidsCache,
+	bidderName string,
+	bid openrtb2.Bid,
+	warnings []openrtb_ext.ExtBidderMessage) {
+	bidBytes, err := bidsHolder.GetBid(bidderName, bid.ID, bid)
+	if err != nil {
+		warnings = append(warnings, createWarning(err.Error()))
+
+	}
+	bidSplit := strings.Split(path, "seatbid.bid.")
+	value, err := splitAndGet(bidSplit[1], bidBytes, pathDelimiter)
+	if err != nil {
+		warnings = append(warnings, createWarning(err.Error()))
+
+	}
+	targetingData[key] = value
+}
```

Comment:
IMHO it might be easier to read the method by moving the parameters to a single line. I've also noticed there is no return values which increases the number of arguments. How do you feel about:

```
func getValueFromSeatBidBid(path string, targetingData map[string]string, bidsHolder bidsCache, bidderName string, bid openrtb2.Bid) error {
	bidBytes, err := bidsHolder.GetBid(bidderName, bid.ID, bid)
	if err != nil {
		return createWarning(err.Error())

	}

	path = strings.TrimPrefix(path, "seatbid.bid.")
	value, err := splitAndGet(path, bidBytes, pathDelimiter)
	if err != nil {
		return createWarning(err.Error())

	}

	return value
}
```

where the call site now looks like:

```
if v, err := getValueFromSeatBidBid(path, targetingData, bidsHolder, bidderName, bid); err == nil {
	targetingData[key] = v
} else {
	warnings = append(warnings, err)
}
```

Such that the responsibility of setting the response and collecting errors is moved up a level. Same comment applies to the other getValue methods.

Reply:
This is a great point! How I didn't see it while coding (I wanted to hide all those appends and set to map statements)... It's even better, error check and value set is only required once for all "getValue" functions at the end of `processResponseTargetingData`